source/world_builder/interface.cc

```diff
@@ -0,0 +1,1082 @@
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <aspect/global.h>
+#include <aspect/world_builder/interface.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/std_cxx11/tuple.h>
+#include <deal.II/base/utilities.h>
+#include <list>
+#include <sstream>
+#include <regex>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+
+namespace aspect
+{
+  namespace WorldBuilder
+  {
+    template <int dim>
+    Interface<dim>::~Interface ()
+    {}
+
+
+    template <int dim>
+    void
+    Interface<dim>::initialize ()
+    {}
+
+    template <int dim>
+    void
+    Interface<dim>::
+    declare_parameters (WorldBuilderParameterHandler &)
+    {}
+
+
+    template <int dim>
+    void
+    Interface<dim>::parse_parameters (dealii::ParameterHandler &)
+    {}
+
+
+
+
+    // -------------------------------- Deal with registering world generator models and automating
+    // -------------------------------- their setup and selection at run time
+
+    namespace
+    {
+      std_cxx11::tuple
+      <void *,
+      void *,
+      internal::Plugins::PluginList<Interface<2> >,
+      internal::Plugins::PluginList<Interface<3> > > registered_plugins;
+    }
+
+
+
+    template <int dim>
+    void
+    Manager<dim>::register_world_builder (const std::string &name,
+                                          const std::string &description,
+                                          void (*declare_parameters_function) (ParameterHandler &),
+                                          Interface<dim> *(*factory_function) ())
+    {
+      std_cxx11::get<dim>(registered_plugins).register_plugin (name,
+                                                               description,
+                                                               declare_parameters_function,
+                                                               factory_function);
+    }
+
+    namespace
+    {
+      std::string
+      mangle (const std::string &s)
+      {
+        std::string u;
+
+        // reserve the minimum number of characters we will need. it may
+        // be more but this is the least we can do
+        u.reserve (s.size());
+
+        // see if the name is special and if so mangle the whole thing
+        const bool mangle_whole_string = (s == "value");
```

Comment:
I don't understand this, if s==value, couldn't you just return the mangled name of "value"?